Codebase for checks

DanielC-N opened the pull request:
features added : 
- possibility to get a source/reference text
- Reference link : when clicked, it scrolls directly to the correct verse and highlight it for 3 seconds. 
- checks :
  - [X] : integrity checks : Checks for missing, duplicated, or out-of-order chapter/verse numbers.
  - [X] : verse stats (with a reference/source text) : Detects short, long, or empty verses based on length comparison between source and target.
  - [X] : unmatched punctuation : Detects unmatched punctuation pairs across verses (e.g., parentheses).
  - [X] : repeated words and whitespaces : Detects consecutive repeated words and excessive whitespace in verses.

TODO : 
~~Quotations detection report : check if the quote in the source language exists in the target language.~~
- [x] : Numbers : check if the numbers from the source translation are correctly reported in the target translation

Commit:
new clean branch for checks

Commit:
bible-check works within scribe env

Commit:
updated bible-checker and enhanced front end for the checks

Commit:
checks OK : verse_stats, integrity_check, repeated_words_whitespace, unmatched_punctuation and a way to automatically scroll to the correct chapter/verse reference with a highlight

Commit:
changed the way the verse are highlighted

selvabenz commented:
1. Instead of no content available display choose a check from the checks dropdown 
2. Do need to load the source every time i open the project or is the selected source Bible saved? how do i know which source is loaded
3. Can we keep the highlighted verse till the verse corrects it or can we retain the highlight for a longer period?
4. Can we stick the checks heading while scrolling the mistakes?
~~5. It says too long but detecting only the part of it MAT:6:8~~ updated PR has the solution
6. Not taking the English ULT as  a source
7. Non of the IRV Bibles are working as source Bible for the checks 
8. Keep a provision for ignoring the checks with need the repeated words
9. Finding the probelem in the Integrity Check but not taking it to the right reference as mentioned in the checks (When the verse number is not there, it is taken to the next chapter verse)
10. Display the name of the source Bible loaded on the check panel
11. Immediate update is not happening in the check after the correction

DanielC-N commented:
1 : Instead of no content available display choose a check from the checks dropdown :green_circle: :  `this message is correctly displayed when no checks are selected.`
3 : Can we keep the highlighted verse till the verse corrects it or can we retain the highlight for a longer period? :green_circle:  `Now it last 5 seconds`
5 : Not taking the English ULT as a source :question: `it does but it takes a long time because of the usfm => usj conversion`
6 : Non of the IRV Bibles are working as source Bible for the checks  :question: `same here.`
8:  Finding the probelem in the Integrity Check but not taking it to the right reference as mentioned in the checks (When the verse number is not there, it is taken to the next chapter verse)   :green_circle:  `Problem solved. When the verse doesn't exist, it takes you to the precedent verse, if it can't find it, it'll scroll to the chapter at least.`
9 : Display the name of the source Bible loaded on the check panel  :green_circle: 
10 : Immediate update is not happening in the check after the correction  :green_circle:

Commit:
fix for : https://github.com/bible-technology/scribe-scripture-editor/pull/389#issuecomment-2563698708

selvabenz commented:
Thank you @DanielC-N for the fixes. The following issues found during the testing.

1. In the first attempt of the checks the statistical verse is not shown/loading, after navigating to someother book then it shows the statistical verse 
![image](https://github.com/user-attachments/assets/a272a953-2e95-45ec-bf88-6cc0b4ccebf6)

1. Short verse shown as too long verse in the checks result
![image](https://github.com/user-attachments/assets/80300d8a-885b-456f-9cb5-5740c69d6003)
![image](https://github.com/user-attachments/assets/931b588d-9c73-4a35-947d-68fcc177993c)

3. Integrity and repeated verse checks are not scrollable 
![image](https://github.com/user-attachments/assets/857ca881-ac28-4aeb-be2c-e887af9f55ab)

5. Verse statistics: the reference link detects correctly in the checks, but in the editor it points to a different verse
![image](https://github.com/user-attachments/assets/410b5624-4b71-4607-9761-9dc4d9932b96)

7. Picking only half of the verse and it takes to the wrong verse number
![image](https://github.com/user-attachments/assets/dea89439-2ad3-4306-aaac-a9d222374ece)
![image](https://github.com/user-attachments/assets/467f732b-6add-4336-8e87-783b599d58cf)

9. Can we put the verse number missing as a separate category? In the description of the checks it says empty verse
![image](https://github.com/user-attachments/assets/00de9d10-bb7d-42b2-bb87-76fb3fec4778)

DanielC-N commented:
1. In the first attempt of the checks the statistical verse is not shown/loading, after navigating to someother book then it shows the statistical verse  :green_circle:  : Fixed.
2. Short verse shown as too long verse in the checks result :green_circle:  : Fixed.
3. Integrity and repeated verse checks are not scrollable  :green_circle:  : Fixed.
4. Verse statistics: the reference link detects correctly in the checks, but in the editor it points to a different verse  :green_circle:  : Fixed.
5. Picking only half of the verse and it takes to the wrong verse number :green_circle:  : Fixed.
6. Can we put the verse number missing as a separate category? In the description of the checks it says empty verse :orange_circle: : tbd in the `bible-checker` package.

Commit:
fix for https://github.com/bible-technology/scribe-scripture-editor/pull/389#issuecomment-2564158269

selvabenz commented:
Thank you @DanielC-N for the fixes. please check the mentioned issues

1. In the integrity check the reference link is taking me to the particular chapter not to the verse mentioned in the link
![image](https://github.com/user-attachments/assets/dda3f1b1-293e-46ff-84a2-574a1b595e11)

3. The repeated verses and the excessive whitespace section not scrolling fully
![image](https://github.com/user-attachments/assets/279cd378-1dc8-4a16-abf4-9e4880b6aaf8)

5. Short verse shown as too long verse in the checks result : **Not Fixed**. it still shown as too long verse please check the screenshot
![image](https://github.com/user-attachments/assets/5ad33a1d-1114-438d-93a0-01eac7343322)

7. If the verse is missing or if the problem occurs in the verse before the heading then it takes the heading along with the verse 
![image](https://github.com/user-attachments/assets/41408191-c509-4d96-947a-6fb7019cdc39)
![image](https://github.com/user-attachments/assets/86c8c85c-ce9c-411f-ae69-9dcdf218d151)

9. Not important: the given time delay is displayed together. we can close the highlight of the previous when the latest is clicked 
![image](https://github.com/user-attachments/assets/f28000cc-5483-408e-b20b-12a602ff9be4)

DanielC-N commented:
1. In the integrity check the reference link is taking me to the particular chapter not to the verse mentioned in the link. :orange_circle:  : to discuss.
2. The repeated verses and the excessive whitespace section not scrolling fully :green_circle:  :  I've found a reliable fix for the scrolling problem AND now the headers of each checks stick to the top when you're scrolling!
3. Short verse shown as too long verse in the checks result : Not Fixed. it still shown as too long verse please check the screenshot :orange_circle: : TBD in the next update of the package `bible-checker`
4. If the verse is missing or if the problem occurs in the verse before the heading then it takes the heading along with the verse :green_circle:  : the link doesn't highlight anymore the headers. Is it ok like this?
![do_not_highlight_headers](https://github.com/user-attachments/assets/0ae12a58-8472-4382-923d-64174f3515e3)

5.Not important: the given time delay is displayed together. we can close the highlight of the previous when the latest is clicked :orange_circle:  : this would be not efficient at all to code something like that ... I'll do that if we have time.

Commit:
some fixes for scrolling problem and highlight problems. See : https://github.com/bible-technology/scribe-scripture-editor/pull/389#issuecomment-2565477073

Commit:
minor fix for scrolling

Commit:
removed a 'console log'

Commit:
it's now highlighting the verse also in the reference if found

Commit:
updated bible-checker

selvabenz commented:
@DanielC-N , Could you check this error?

![image](https://github.com/user-attachments/assets/417be901-c95e-4fcd-abe2-ad7d5263fac8)

Commit:
fix for : https://github.com/bible-technology/scribe-scripture-editor/pull/389#issuecomment-2566331519

DanielC-N commented:
> Can we stick the checks heading while scrolling the mistakes?  

:green_circle: : now the headings stick to the top as you scroll

>Do need to load the source every time i open the project or is the selected source Bible saved? how do i know which source is loaded  

:green_circle: : now it shows wich resource has been selected.

> Short verse shown as too long verse in the checks result : Not Fixed. it still shown as too long verse.  

:green_circle:  :  now short and long verses should be displayed correctly

> Can we put the verse number missing as a separate category? In the description of the checks it says empty verse  

:green_circle:  : now the missing verses are displayed in a separated category

> Non of the IRV Bibles are working as source Bible for the checks (as well as some other resources  

:orange_circle:  : we've discussed with @samueljd , it's a problem of convertion from USFM to USJ. It's in progress and should be fixed with Sam. I'll make sure the app doesn't crash.

Commit:
updated bible-check with number checking

DanielC-N commented:
last commit fixes/improvments : 
- FIX : add verse/chapter popup is no longer showing behind the check bar :green_circle: 
- FIX : the project reference name doesn't show the long version with the random numbers :green_circle: 
- UPDATE : now the source verse is shown in the check menu :green_circle: 

Known bug : 
- some times the reference doesn't scroll correctly to the correct sentence in the reference text panel. :orange_circle:

Commit:
update for few fixes : https://github.com/bible-technology/scribe-scripture-editor/pull/389#issuecomment-2571276002

Commit:
final fixes for bible-checks
